prime-router/src/main/kotlin/fhirengine/utils/HL7Reader.kt

```diff
@@ -18,33 +23,122 @@ import java.util.Date
             } catch (e: Hl7InputStreamMessageStringIterator.ParseFailureError) {
                 logHL7ParseFailure(e)
             }
+        } else {
+            val validationContext = ValidationContextFactory.noValidation()
+            var parseError: Hl7InputStreamMessageStringIterator.ParseFailureError? = null
+            run modelLoop@{
+                messageModelsToTry.forEach { model ->
+                    val context = DefaultHapiContext(CanonicalModelClassFactory(model))
+                    context.validationContext = validationContext
+                    try {
+                        val iterator = Hl7InputStreamMessageIterator(rawMessage.byteInputStream(), context)
+                        while (iterator.hasNext()) {
+                            messages.add(iterator.next())
+                        }
+                    } catch (e: Hl7InputStreamMessageStringIterator.ParseFailureError) {
+                        parseError = e
+                    }
+
+                    if (messages.isNotEmpty()) {
+                        // Don't try other message models if we were able to parse
+                        return@modelLoop
+                    }
+                }
+            }
 
-            if (messages.isEmpty() && !actionLogger.hasErrors()) {
-                actionLogger.error(InvalidReportMessage("Unable to find HL7 messages in provided data."))
+            // This is a known kotlin bug where the compiler does not think parseError can be smart-casted because
+            // it is operated on in multiple forEach closures, the solution is to just reassign
+            // https://youtrack.jetbrains.com/issue/KT-19446/False-positive-Smart-cast-to-Foo-is-impossible-due-to-same-variable-names-in-different-closures
+            val parseErrorToLog = parseError
+            if (parseErrorToLog != null) {
+                // Only log a parse failure if all the model classes have been tried and no messages have been parsed
+                logHL7ParseFailure(parseErrorToLog, messages.isEmpty())
             }
         }
+
+        if (messages.isEmpty() && !actionLogger.hasErrors()) {
+            actionLogger.error(InvalidReportMessage("Unable to find HL7 messages in provided data."))
+        }
+
         return messages
     }
 
+    /**
+     * Extracts the message type from the MSH segment and returns the list of message models to use to
+     * try to parse the messages.
+     *
+     * This function assumes all the message types will be the same if this is a HL7 batch.
+     */
+    private fun getMessageModelClasses(rawMessage: String): List<Class<out AbstractMessage>> {
+        val iterator = Hl7InputStreamMessageIterator(rawMessage.byteInputStream())
+        if (iterator.hasNext()) {
+            try {
+                val firstMessage = iterator.next()
+                return when (val messageType = getMessageType(firstMessage)) {
```

Comment:
Thoughts on adding a function `getMessageStructure` to use here in case we ever need an `ORU` which isn't `R01`? Could also be a backlog item, this seems safe enough for now.

Reply:
I